sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -277,6 +278,31 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
     result.copy(output = newOutput)
   }
 
+  private def getDirectoryPathSeq(rootPath: Path): Seq[String] = {
+    val enableSupportSubDirectories =
+      sparkSession.conf.getOption("hive.mapred.supports.subdirectories")
+
+    if (enableSupportSubDirectories.isDefined && enableSupportSubDirectories.get.toBoolean) {
+      val fs = rootPath.getFileSystem(sparkSession.sessionState.newHadoopConf())
+      val paths = new scala.collection.mutable.ListBuffer[String]
+
+      val checkingQueue = new mutable.Queue[Path]()
+      checkingQueue.enqueue(rootPath)
+      while (!checkingQueue.isEmpty) {
+        val path = checkingQueue.dequeue()
+        paths.append(path.toString)
+        fs.listStatus(path).foreach(fileStatus => {
+          if (fileStatus.isDirectory) {
+            checkingQueue.enqueue(fileStatus.getPath)
+          }
+        })
+      }
+      paths
```

Comment:
Why not using the existing `SparkHadoopUtil.listLeafDirStatuses(fs, rootPath)`?

Reply:
Didn't notice the function before, will change it. Thanks.